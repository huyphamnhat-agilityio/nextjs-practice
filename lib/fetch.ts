export const fetchApi = async <T>(url: string, options?: RequestInit) => {
  const response = await fetch(url, {
    method: options?.method || "GET",
    headers: options?.headers || {
      "Content-Type": "application/json",
    },
    body: options?.body,
    ...options,
  });

  if (response.ok) return response.json() as T;

  const errorMessage = `${response.status}: ${await response.text()}`;

  throw errorMessage;
};
