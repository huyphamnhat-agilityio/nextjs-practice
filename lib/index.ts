export * from "./fetch";
export {
  getProducts,
  markProduct,
  mutateProduct,
  deleteProduct,
} from "./product";
export { subscribe } from "./subscribe";
export { buildRedirectPath } from "./redirect";
