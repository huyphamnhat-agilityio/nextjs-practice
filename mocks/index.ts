export * from "./category";
export * from "./product";
export * from "./fetch";
export * from "./avatar";
export * from "./image";
