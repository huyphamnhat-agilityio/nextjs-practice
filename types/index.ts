export * from "./navbar";
export * from "./product";
export * from "./category";
