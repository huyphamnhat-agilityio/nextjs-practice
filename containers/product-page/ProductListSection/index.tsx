"use client";
import { PRODUCT_MESSAGES, TOAST_SECTION, TOAST_TYPE } from "@/constants";
import { ToastType } from "@/types";
import { usePathname, useRouter, useSearchParams } from "next/navigation";
import { ReactNode, useEffect } from "react";
import toast from "react-hot-toast";

const ProductListSection = ({ children }: { children: ReactNode }) => {
  const searchParams = useSearchParams();
  const pathname = usePathname();
  const router = useRouter();

  const toastType = searchParams.get("toastType");
  const toastSection = searchParams.get("toastSection");
  const message = searchParams.get("message");

  useEffect(() => {
    if (toastSection === TOAST_SECTION.PRODUCT_LIST_SECTION) {
      toastType === TOAST_TYPE.SUCCESS
        ? toast.success(message)
        : toast.error(message);

      const params = new URLSearchParams(searchParams.toString());

      params.delete("toastType");
      params.delete("message");
      params.delete("toastSection");

      router.push(`${pathname}?${params.toString()}`, { scroll: false });
    }
  }, [message, pathname, router, searchParams, toastSection, toastType]);

  return <div>{children}</div>;
};

export default ProductListSection;
