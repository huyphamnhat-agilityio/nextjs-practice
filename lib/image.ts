// Types
import { ImgBBResponse } from "@/types";

import { fetchApi } from "./fetch";
import { headers } from "next/headers";

export const uploadAndGetImageUrl = async (image: File) => {
  console.log("image", image);
  const imageFormData = new FormData();

  imageFormData.append("image", image);

  const options = {
    method: "POST",
    body: imageFormData,
    headers: undefined,
  };

  const response = await fetchApi<ImgBBResponse>(
    `${process.env.UPLOAD_IMAGE_API}?key=${process.env.IMGBB_API_KEY}`,
    {
      method: "POST",
      body: imageFormData,
      headers: undefined,
    },
  );

  return response.data.url;
};
