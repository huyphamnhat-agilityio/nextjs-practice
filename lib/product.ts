"use server";
import { fetchApi } from "./fetch";

// Types
import { Pagination, Product } from "@/types";

// Constants
import { RESOURCES } from "@/constants";
import { revalidatePath } from "next/cache";

export type GetProductProps = {
  page: number;
  limit: number;
  query?: string;
};
export const getProducts = async ({
  page,
  limit,
  query = "",
}: GetProductProps) => {
  const options = {
    publicDataUrl: `${process.env.MOCK_API}/${RESOURCES.PRODUCT}?title=${query}`,
    page,
    limit,
  };

  try {
    const products = await fetchApi<Pagination<Product>>(
      process.env.PAGINATION_API,
      {
        method: "POST",
        body: JSON.stringify(options),
        next: {
          revalidate: 300,
        },
      },
    );

    return products;
  } catch (error) {
    if (error.status === 404 || error.status === 400) return undefined;
    throw new Error(error.message);
  }
};

export const markProduct = async (data: Product) => {
  try {
    const response = await fetchApi<Product>(
      `${process.env.MOCK_API}/${RESOURCES.PRODUCT}/${data.id}`,
      {
        method: "PUT",
        body: JSON.stringify({ ...data, isFavorited: !data.isFavorited }),
      },
    );
    revalidatePath("/", "layout");

    return response;
  } catch (error) {
    throw error;
  }
};

export const createProduct = async (data: FormData) => {
  console.log(data.get("description"));
};
