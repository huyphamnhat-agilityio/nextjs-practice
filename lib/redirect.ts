import { ToastSection, ToastType } from "@/types";

export type BuildRedirectPathProps = {
  pathname: string;
  type: ToastType;
  section: ToastSection;
  message: string;
};
export const buildRedirectPath = ({
  pathname,
  type,
  section,
  message,
}: BuildRedirectPathProps) => {
  const delimiter = pathname[pathname.length - 1] === "?" ? "" : "&";
  return `${pathname}${delimiter}toastType=${type}&toastSection=${section}&message=${message}`;
};
