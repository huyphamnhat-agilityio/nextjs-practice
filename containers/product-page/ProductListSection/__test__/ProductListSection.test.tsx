import { render, screen } from "@testing-library/react";
import { usePathname, useRouter, useSearchParams } from "next/navigation";
import { act } from "react";

// Components
import ProductListSection from "..";

// Services
import { buildRedirectPathWithToast } from "@/lib";

// Constants
import {
  DESTINATION,
  PRODUCT_MESSAGES,
  TOAST_ACTION,
  TOAST_SECTION,
  TOAST_TYPE,
} from "@/constants";

// Types
import { Pagination, Product } from "@/types";
import { MOCK_PRODUCTS } from "@/mocks";
import userEvent from "@testing-library/user-event";

jest.mock("next/navigation", () => ({
  usePathname: jest.fn(),
  useSearchParams: jest.fn(),
  useRouter: jest.fn(),
}));

jest.mock("react-dom", () => ({
  ...jest.requireActual("react-dom"),
  useFormStatus: jest.fn(),
  useFormState: () => [{}, () => {}],
}));

describe("ProductListSection test cases", () => {
  const mockPush = jest.fn();
  const mockReplace = jest.fn();
  const mockUseRouter = useRouter as jest.Mock;
  const mockUsePathname = usePathname as jest.Mock;
  const mockUseSearchParams = useSearchParams as jest.Mock;
  const mockProducts: Pagination<Product> = {
    data: MOCK_PRODUCTS,
    hasNextPage: true,
    hasPrevPage: false,
    limit: 3,
    page: 1,
    totalItems: MOCK_PRODUCTS.length,
    totalPages: 2,
    nextPage: 2,
    prevPage: null,
  };

  beforeEach(() => {
    mockUseRouter.mockReturnValue({ replace: mockReplace, push: mockPush });
    mockUsePathname.mockReturnValue(DESTINATION.PRODUCT);
    mockUseSearchParams.mockReturnValue(new URLSearchParams());
  });

  const setup = ({ products }: { products: Pagination<Product> }) =>
    render(<ProductListSection products={products} />);
  it("should render correctly", async () => {
    const { asFragment } = await act(() => setup({ products: mockProducts }));

    expect(asFragment()).toMatchSnapshot();
  });

  it("shoulde able to show success toast corresponding to the query params", async () => {
    let mockParams = buildRedirectPathWithToast({
      pathname: DESTINATION.PRODUCT,
      type: TOAST_TYPE.SUCCESS,
      section: TOAST_SECTION.PRODUCT_LIST_SECTION,
      action: TOAST_ACTION.CONFIRM,
      message: PRODUCT_MESSAGES.SUCCESS.DELETE,
    });

    mockUseSearchParams.mockReturnValue(new URLSearchParams(mockParams));

    const { asFragment } = await act(() => setup({ products: mockProducts }));

    expect(asFragment()).toMatchSnapshot();
  });

  it("shoulde able to show error toast corresponding to the query params", async () => {
    let mockParams = buildRedirectPathWithToast({
      pathname: DESTINATION.PRODUCT,
      type: TOAST_TYPE.ERROR,
      section: TOAST_SECTION.PRODUCT_LIST_SECTION,
      action: TOAST_ACTION.CONFIRM,
      message: PRODUCT_MESSAGES.ERROR.DELETE,
    });

    mockUseSearchParams.mockReturnValue(new URLSearchParams(mockParams));

    const { asFragment } = await act(() => setup({ products: mockProducts }));

    expect(asFragment()).toMatchSnapshot();
  });

  it("should be able to navigate to the next page", async () => {
    setup({ products: mockProducts });

    const paginationBtn = screen.getByRole("button", {
      name: /pagination item 2/i,
    });

    await userEvent.click(paginationBtn);

    expect(mockReplace).toHaveBeenCalled();
  });
});
