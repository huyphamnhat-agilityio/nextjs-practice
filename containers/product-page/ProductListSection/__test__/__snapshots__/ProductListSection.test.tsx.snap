// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductListSection test cases should render correctly 1`] = `
<DocumentFragment>
  <div />
</DocumentFragment>
`;

exports[`ProductListSection test cases shoulde able to show error toast corresponding to the query params 1`] = `
<DocumentFragment>
  <div />
</DocumentFragment>
`;

exports[`ProductListSection test cases shoulde able to show success toast corresponding to the query params 1`] = `
<DocumentFragment>
  <div />
</DocumentFragment>
`;
